From 0ba765f4974db890dcc7379b2b589f8216eede28 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:22 -0400
Subject: [powerpc] scom: Add support for "reg" property

Message-id: <1410545655-205645-94-git-send-email-dzickus@redhat.com>
Patchwork-id: 94169
O-Subject: [RHEL7 PATCH 093/626] powerpc/scom: Add support for "reg" property
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 5f33af4c0059255bcbf82a98a3789a01171b72e5
Author: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Date:   Thu Aug 29 16:56:16 2013 +1000

    powerpc/scom: Add support for "reg" property

    When devices are direct children of a scom controller node, they
    should be able to use the normal "reg" property instead of "scom-reg".

    In that case, they also use #address-cells rather than #scom-cells
    to indicate the size of an entry.

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/sysdev/scom.c b/arch/powerpc/sysdev/scom.c
index 10f1d9e..413622d 100644
--- a/arch/powerpc/sysdev/scom.c
+++ b/arch/powerpc/sysdev/scom.c
@@ -53,7 +53,7 @@ scom_map_t scom_map_device(struct device_node *dev, int index)
 {
  struct device_node *parent;
  unsigned int cells, size;
- const u32 *prop;
+ const __be32 *prop, *sprop;
  u64 reg, cnt;
  scom_map_t ret;
 
@@ -62,12 +62,24 @@ scom_map_t scom_map_device(struct device_node *dev, int index)
  if (parent == NULL)
   return 0;
 
- prop = of_get_property(parent, "#scom-cells", NULL);
- cells = prop ? *prop : 1;
-
+ /*
+  * We support "scom-reg" properties for adding scom registers
+  * to a random device-tree node with an explicit scom-parent
+  *
+  * We also support the simple "reg" property if the device is
+  * a direct child of a scom controller.
+  *
+  * In case both exist, "scom-reg" takes precedence.
+  */
  prop = of_get_property(dev, "scom-reg", &size);
+ sprop = of_get_property(parent, "#scom-cells", NULL);
+ if (!prop && parent == dev->parent) {
+  prop = of_get_property(dev, "reg", &size);
+  sprop = of_get_property(parent, "#address-cells", NULL);
+ }
  if (!prop)
-  return 0;
+  return NULL;
+ cells = sprop ? be32_to_cpup(sprop) : 1;
  size >>= 2;
 
  if (index >= (size / (2*cells)))
-- 
1.7.1